### DISCUSSION TITLE
How to display array format of a parameter in url when I pass it to query?

### DISCUSSION THREAD

**Discussion by alamenai** [vercel/next.js#51198] [Help]
## Task
I'm working on NextJS 12 project and I want to display an array format ( ex: `/path?id=1&rooms=[1,2]`) in the URL.

## Code

```javascript
router.replace({
					pathname: '/panel_placement',
					query: {
						address_id: '11842008',
						roofs:`[${selectedRoofs.join(',')}]`, //here
						lang,
					},
				});
```

## Expected

The format should be like this :

 `/path?address_id=1235&roofs=[1,2]&lang=en`

## Current

However, when I pass the array I get an encoded result : 

 `/path?address_id=1235&roofs=%5B1%2C2%5D&lang=en`

**Comment by devjiwonchoi** [comment]
You can decode the given `URL` using `decodeURIComponent` since it is an encoded URL.

```js
const url = '/path?address_id=1235&roofs=%5B1%2C2%5D&lang=en'
console.log(decodeURIComponent(url)) // Output: /path?address_id=1235&roofs=[1,2]&lang=en
```

## Reference
- [decodeURIComponent()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent)
- [URL Encoding Reference](https://www.w3schools.com/tags/ref_urlencode.ASP)

**Comment by icyJoseph** [comment]
Characters such as `[` or `]` are not necessarily allowed in the URL, without encoding. What browsers have come to do is to please the user, is to just show them, but they work with the encoded values. Servers also recognize these and can parse them properly into a list.

So I think yo have to do this manually: `router.replace("/?tags=[1,2]");`

**Reply by alamenai** [reply]
In my case:

```javascript

moveToNextPage = () =>{

const query = {
			address_id: router.query.address_id,
			selected_roofs: selectedRoofs.join(''),
			stride_x: panelWidth,
			stride_y: panelHeight,
			border_space: borderSpacing,
			panel_space: panelSpacing,
		};

router.replace({
				pathname: '/panel_placement',
				query,
			});

}
```

This code produces this URL:

**http://localhost:3000/panel_placement?address_id=11842008&selected_roofs=12&stride_x=1.1&stride_y=1.7&border_space=0.05&panel_space=0.07**

I want this format :

http://localhost:3000/panel_placement?address_id=11842008&selected_roofs=[1,2]&stride_x=1.1&stride_y=1.7&border_space=0.05&panel_space=0.07

What should I change to get this format where `selected_roofs=12` becomes `selected_roofs=[1,2]`.

**Reply by icyJoseph** [reply]
Something like:

```js
         const query = {
          address_id: 11842000,
          selected_roofs: [1, 2],
          stride_x: 1.1,
          stride_y: 1.7,
          border_space: 0.05,
          panel_space: 0.07,
        };

        const asString = Object.entries(query).reduce((prev, [key, value]) => {
          const pair = `${key}=${
            Array.isArray(value) ? `[${value}]` : `${value}`
          }`;

          return `${prev}&${pair}`;
        }, ""); // do not start with "?"

        // add the "?"
        router.replace(`/panel_placement?${asString}`);
```

**Comment by alamenai** [comment]
@devjiwonchoi , my goal is for the user can change the roof array directly from the URL itself.
I know it is impossible to encode and decode the URL but how to let the user modify the URL query parameters ( arrays ) in the URL directly?

**Reply by icyJoseph** [reply]
What do you mean with:

>  let the user modify the URL query parameters 

In what way would they do that? The browser allows it, and as I said in my other answer, you can just do:

```js
import { useRouter } from "next/router";

export default function Home() {
  const router = useRouter();
  return (
    <button
      onClick={() => {
        router.replace("/?tags=[1,2]");
      }}
    >
      Replace
    </button>
  );
}
```

And that'll show the list in the URL, but it is just the browser setting up a facade to visually please the user.

https://github.com/vercel/next.js/assets/21013447/68d94983-0bde-4984-aeba-434144edbf59

If you use the `object` way to push/replace, that'll run through the URLSearchParams or URL API, and when casted back to string it is automatically encoded.

```js
const q = {tags: '[0,1,2]'}
const search = new URLSearchParams(q)
search.get("tags") // '[0,1,2]'
search.set("foo", "[9,7,8]")
search.get("foo") // '[9,7,8]'
search.toString() // 'tags=%5B0%2C1%2C2%5D&foo=%5B9%2C7%2C8%5D'
```

Or

```js
let url = new URL('/foo?q=[1,2,3]', window.location.origin)
url.search // '?q=[1,2,3]'
url.searchParams.toString() // 'q=%5B1%2C2%2C3%5D'
```
